lib/bap_disasm/bap_disasm_reconstructor.ml

```diff
@@ -46,39 +46,68 @@ let is_start roots cfg blk =
   Set.mem roots (Block.addr blk) ||
   Cfg.Node.degree ~dir:`In blk cfg = 0
 
-let entries_of_block cfg roots entries blk =
+let is_fall e = Cfg.Edge.label e = `Fall
+let is_call b = Insn.(is call) (Block.terminator b)
+
+let entries_of_block cfg roots blk =
   let entries =
-    if is_start roots cfg blk then Set.add entries blk
-    else entries in
-  let term = Block.terminator blk in
-  if Insn.(is call) term then
+    if is_start roots cfg blk then Block.Set.singleton blk
+    else Block.Set.empty in
+  if is_call blk then
     Seq.fold ~init:entries (Cfg.Node.outputs blk cfg)
       ~f:(fun entries e ->
-          if Cfg.Edge.label e <> `Fall then
-            Set.add entries (Cfg.Edge.dst e)
-          else entries)
+          if is_fall e then entries
+          else Set.add entries (Cfg.Edge.dst e))
   else entries
 
-let collect_entries cfg roots =
-  let roots = Addr.Set.of_list roots in
-  Seq.fold (Cfg.nodes cfg) ~init:Block.Set.empty
-    ~f:(entries_of_block cfg roots)
+let terminator_addr blk =
+  List.rev (Block.insns blk) |>
+  List.hd_exn |>
+  fst |>
+  Memory.min_addr
+
+let maybe_external blk cfg =
+  let deg = Cfg.Node.degree ~dir:`Out blk cfg in
+  deg = 0 ||
+  (deg = 1 && is_fall (Seq.hd_exn (Cfg.Node.outputs blk cfg)))
+
+let add_callees syms name cfg blk =
+  if is_call blk then
+    let call_addr = terminator_addr blk in
+    if maybe_external blk cfg then
```

Comment:
We're here assuming that any unresolved call is an external call. I think that it is too overly, we need somehow to limit the set of presumed externals. So, we will consider the call to be external if it has a resolved name, i.e., something that is not `sub_XXX`. Of course, we won't look at the name for that, instead we will extend the symbolizer interface with a new function `is_resolved` that will return true if a symbolizer returned `Some name`.